src/NUnitFramework/framework/Internal/Commands/TimeoutCommand.cs

```diff
@@ -48,11 +50,15 @@ public class TimeoutCommand : BeforeAndAfterTestCommand
         /// </summary>
         /// <param name="innerCommand">The inner command</param>
         /// <param name="timeout">Timeout value</param>
-        public TimeoutCommand(TestCommand innerCommand, int timeout) : base(innerCommand)
+        /// <param name="debugger">An <see cref="IDebugger"/> instance</param>
+        internal TimeoutCommand(TestCommand innerCommand, int timeout, IDebugger debugger) : base(innerCommand)
```

Comment:
This is a problem. The TimeoutCommand type is not usable at all unless its constructor is usable. We may as well make the whole type internal. @nunit/framework-team Since this is under the Internal namespace, are you okay with that?

Reply:
@jnm2 (I know I'm not on the team but I hope this may be useful)

The design principle behind custom attributes was that users would implement interfaces and not reuse code. Based on that principle, I would say this is __not__ breaking.

OTOH the framework has accepted bugs in recent years where people re-used such classes, even though they were not designed to be reused. And we know from experience that some people do re-use classes rather than just implement the interfaces. So your guess is as good as mine... maybe better.

Of course, this is just one instance of a big disagreement we've kicked around for a few years: is it OK to break an accessible class if people were not "supposed" to use it by some definition or another?

Anyway, it's out of my wheelhouse now, but that's how I see the choice.

Reply:
That's a good summary. If we think people may be using this class directly, we will have to preserve the `public TimeoutCommand(TestCommand innerCommand, int timeout)` constructor since I can't think of a way this class could be used from the outside without calling this constructor. If we don't think people may be using this class directly, or we are willing to break their code, then there's no reason to keep the class public.

For a member like this constructor which is essential to the use of the class itself, the member needs to be as accessible as the whole class. To do otherwise is a design smell.

Reply:
Ahhh, this old chestnut. 🙂 

My view: TimeoutCommand is new-ish (2017, https://github.com/nunit/nunit/pull/2106) - and has been in the Internal namespace its entire existence. I agree the Internal namespace isn't anywhere near an ideal solution, but I do think it's a policy that we can stand by. If we go the other way and make the new DebuggerProxy classes public, we're just making the overall problem worse - so I think we should take a risk on making TimeoutCommand internal, as the lesser evil.

@jnm2 /@CharliePoole /@zdenek- : Can we make come to a decision between the four of us,  so this doesn't end up stuck again and we can get it merged? 😄 Who wants to disagree with me? 😁

Reply:
Looks like GitHub's autocomplete messed up on you there. /cc @zdenek-jelinek

I don't disagree. It's worth pointing out that we can add this internal constructor and leave the existing public constructor in place. We wouldn't be talking about making IDebugger public necessarily.

Reply:
Maybe that's the best solution - it's low-cost, and zero risk.

Reply:
If we plan to clean up in v4 by making the class internal, we could also signal that now with an obsoletion message.

Reply:
@ChrisMaddock As a user of the framework, it's a don't care for me. I think your point of view makes sense and I'd support it if I had a vote, which I don't. Sorry, but as we discussed privately, I take formal roles very seriously.

Reply:
I agree - although I think that could be applied much more widely across a lot of classes in this area. Perhaps out of scope for this PR, and a separate project? 🙂

Reply:
Yes. So I have to say I want to make the class internal, but it's hard to argue with low cost and zero risk and I can't disagree with that. So what we're asking @zdenek-jelinek to do is to restore the original public constructor with its signature unchanged. It would call the internal constructor so that it would continue to work as it had previously for users depending on that constructor signature.

Reply:
Just one more question from my side: How does an internal consumer (i.e. person working on NUnit framework) know which constructor to use? Should I [Obsolete("Will be removed in the next major release")] the public ctor, or just leave it as it were?

Reply:
@zdenek-jelinek Good point. Either that or an XML doc comment, but let's use do the obsoletion warning.